
def printHelp() {
  log.info"""
  Usage:
    nextflow run BCCDC-PHL/mpxv-artic-nf -profile conda --prefix [prefix] [workflow-options]

  Description:
    Turn Illumina monkeypox virus (mpxv) sequencing reads generated by ARTIC tiling amplification protocols into consensus sequences.

    All options set via CLI can be set in conf directory

  Nextflow arguments (single DASH):
    -profile                      Allowed values: conda
 
  Workflow options:
    Mandatory:
      --prefix                    A (unique) string prefix for output files.
                                  Sequencing run name is a good choice e.g DDMMYY_MACHINEID_RUN_FLOWCELLID.
      --directory                 Path to a directory containing paired-end Illumina reads. 
                                  Reads will be found and paired RECURSIVELY beneath this directory.
    Optional:
      --outdir                    Output directory (Default: ./results)

      --scheme_version            Primer scheme version (Default: 'INRB/V3')
      --scheme_name               Scheme name (Default: 'MPXV')
 
      --keepLen                   Length (bp) of reads to keep after primer trimming (Default: 50)
      --qualThreshold             Sliding window quality threshold for keeping
                                  reads after primer trimming (Default: 20)
      --normalizationTargetDepth  Target depth of coverage to normalize to.
      --normalizationMinDepth     Minimum depth of coverage to normalize to.
      --varMinFreqThreshold       Frequency threshold to call ambiguous variant (Default: 0.25)
      --varFreqThreshold          Frequency threshold for unambiguous variant (Default: 0.75)
      --varMinDepth               Minimum coverage depth to call variant (Default: 10)

  """.stripIndent()
}
